pkg/meta/base.go

```diff
@@ -230,6 +246,90 @@ func (m *baseMeta) checkRoot(inode Ino) Ino {
 	}
 }
 
+func (m *baseMeta) countDirUsage(ctx Context, ino Ino) (space, inodes uint64, err syscall.Errno) {
+	var entries []*Entry
+	if err = m.en.doReaddir(ctx, ino, 1, &entries, -1); err != 0 {
+		return
+	}
+	for _, e := range entries {
+		if string(e.Name) == "." || string(e.Name) == ".." {
+			continue
+		}
+		inodes += 1
+		var newSpace uint64
+		if e.Attr.Typ == TypeFile {
+			newSpace = e.Attr.Length
+		}
+		space += uint64(align4K(newSpace))
+	}
+	return
+}
+
+func (m *baseMeta) GetDirUsage(ctx Context, inode Ino) (space, inodes uint64, err error) {
+	return m.en.doGetDirUsage(ctx, m.checkRoot(inode))
+}
+
+func (m *baseMeta) increDirUsage(ctx Context, ino Ino, space int64, inodes int64) {
+	for i := 0; i < 50; i++ {
+		select {
+		case m.dirUsageEvents <- &dirUsageEvent{ino, space, inodes}:
+			return
+		default:
+			time.Sleep(time.Second * 1)
+		}
+	}
+	logger.Warn("already tried 50 times, dirUsageEvents channel is full, try to update dir usage directly")
+	err := m.en.doIncreDirUsage(ctx, ino, space, inodes)
+	if err != nil {
+		logger.Errorf("update dir usage directly failed: %v", err)
+	}
+}
+
+func (m *baseMeta) increParentUsage(ctx Context, inode, parent Ino, space int64) {
+	if parent > 0 {
+		m.increDirUsage(ctx, parent, space, 0)
+	} else {
+		for p := range m.en.doGetParents(ctx, inode) {
+			m.increDirUsage(ctx, p, space, 0)
+		}
+	}
+}
+
+func (m *baseMeta) dirUsageBatch() {
+	commitChan := make(chan map[Ino]*dirUsageEvent, 16)
+	go func() {
+		for records := range commitChan {
+			for _, r := range records {
+				err := m.en.doIncreDirUsage(WrapContext(context.TODO()), r.ino, r.space, r.inodes)
```

Comment:
We may need to put several inodes within one transaction.

Reply:
I think batch in application is enough, big transactions more likely conflict with other sessions. In this case, maybe we can call `doIncreDirUsage` parallelly as they won't conflict with each other.